source/server/admin/clusters_request.h

```diff
@@ -0,0 +1,128 @@
+
+#include <cstdint>
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/http/codes.h"
+#include "envoy/server/admin.h"
+#include "envoy/server/instance.h"
+
+#include "source/common/buffer/buffer_impl.h"
+#include "source/common/json/json_streamer.h"
+#include "source/server/admin/clusters_params.h"
+
+namespace Envoy {
+namespace Server {
+
+/**
+ * ClustersRequest captures context for a streaming request, implementing the
+ * Admin::Request interface.
+ */
+class ClustersRequest : public Admin::Request {
+public:
+  static constexpr uint64_t DefaultChunkLimit = 2 << 20; // 2 MB
+
+  ClustersRequest(uint64_t chunk_limit, Instance& server, const ClustersParams& params);
+
+  Http::Code start(Http::ResponseHeaderMap& response_headers) override;
+  bool nextChunk(Buffer::Instance& response) override;
+
+protected:
+  uint64_t chunk_limit_{DefaultChunkLimit};
+  Server::Instance& server_;
+  const ClustersParams& params_;
+  std::vector<std::reference_wrapper<const Upstream::Cluster>> clusters_;
+};
+
+/**
+ * TextClustersRequest streams the response using a newline delimited text format.
+ */
+class TextClustersRequest : public ClustersRequest {
+public:
+  TextClustersRequest(uint64_t chunk_limit, Instance& server, const ClustersParams& params);
+  bool nextChunk(Buffer::Instance& response) override;
+
+private:
+  void render(std::reference_wrapper<const Upstream::Cluster> cluster, Buffer::Instance& response);
+
+  static void addOutlierInfo(const std::string& cluster_name,
+                             const Upstream::Outlier::Detector* outlier_detector,
+                             Buffer::Instance& response);
+  static void addCircuitBreakerSettings(const std::string& cluster_name,
+                                        const std::string& priority_str,
+                                        Upstream::ResourceManager& resource_manager,
+                                        Buffer::Instance& response);
+  uint64_t idx_;
+};
+
+/**
+ * ClustersJsonContext holds an Envoy::Json::Streamer and the top-level JSON objects throughout the
+ * duration of a request. When it is destructed, the buffer will terminate the array and object
+ * herein. See the Envoy::Json::Streamer implementation for details.
+ */
+struct ClustersJsonContext {
+  ClustersJsonContext(std::unique_ptr<Json::Streamer> streamer, Buffer::Instance& buffer,
+                      Json::Streamer::MapPtr root_map, Json::Streamer::ArrayPtr clusters)
+      : streamer_(std::move(streamer)), buffer_(buffer), root_map_(std::move(root_map)),
+        clusters_(std::move(clusters)) {}
+  std::unique_ptr<Json::Streamer> streamer_;
+  Buffer::Instance& buffer_;
+  Json::Streamer::MapPtr root_map_;
+  Json::Streamer::ArrayPtr clusters_;
+};
+
+/**
+ * JsonClustersChunkProcessor streams the response and emulates the JSON encoding behavior of
+ * the gRPC clients; specifically, "zero-value" values such as false, 0, empty string, and
+ * empty objects are omitted from the output.
```

Comment:
is this also the behavior of the current impl?

Reply:
Yeah. That's how the proto library marshals json currently.